include/itkSpeckleReducingAnisotropicDiffusionFunction.hxx

```diff
@@ -0,0 +1,185 @@
+SpeckleReducingAnisotropicDiffusionFunction<TImage>::ComputeDiffusionCoefficient(PixelType intensity,
+                                                                                 PixelType averageGradient,
+                                                                                 PixelType laplacian)
+{
+  // Epsilon to avoid divide-by-zero errors
+  const float eps = itk::Math::eps;
+
+  // Compute instantaneous coefficient of variation.
+  // See (35) in Yu, Acton.
+  PixelType numerator = 0.5 * itk::Math::sqr(averageGradient / (intensity + eps)) - itk::Math::sqr(0.25 * laplacian / (intensity + eps));
+  // We only use the real part of q(t) so we discard any values n < 0
+  // that would result in q(t) = sqrt(n/d) = a + b i where b != 0
+  if (numerator < 0)
+  {
+    numerator = 0;
+  }
+
+  PixelType denominator = itk::Math::sqr(1 + 0.25 * laplacian / (intensity + eps));
+
+  // Instantaneous coefficient of variation
+  PixelType q_t = std::sqrt(numerator / denominator);
+
+  // Compute diffusion coefficient.
+  // See (33) in Yu, Acton.
+  PixelRealType C = 1 / (1 + (itk::Math::sqr(q_t) - itk::Math::sqr(m_q0_t)) / (itk::Math::sqr(m_q0_t) * (1 + itk::Math::sqr(m_q0_t)) + eps));
+  return C;
+}
+
+template <typename TImage>
+typename SpeckleReducingAnisotropicDiffusionFunction<TImage>::PixelType
+SpeckleReducingAnisotropicDiffusionFunction<TImage>::ComputeUpdate(const NeighborhoodType & it,
+                                                              void *,
+                                                              const FloatOffsetType &)
+{
+  // Ignore edge pixels. We assume the user has mirror-padded the input image.
+  if (!it.InBounds())
+    return 0;
+
+  unsigned int i;
+
+  std::vector<PixelRealType> c;
+
+  PixelRealType dx;
+  PixelRealType dx_avg;
+  PixelRealType dx2_avg;
+  PixelRealType delta;
+
+  // Create an iterator to walk over a local 2x2 region to generate
+  // diffusion coefficient statistics at each pixel
+  auto    radius = MakeSize(1, 1);
+  RegionType          region;
+  region.SetSize(MakeSize(2,2));
+  region.SetIndex(it.GetIndex());
+  NeighborhoodType coeffIt(radius,it.GetImagePointer(),region);
+
+  while (!coeffIt.IsAtEnd())
+  {
+    // Try to get the diffusion coefficient from the cache
+    // Calculate the centralized derivatives for each dimension.
+    dx = 0;
+    dx_avg = 0;
+    for (i = 0; i < ImageDimension; ++i)
+    {
+      dx = m_InnerProduct(x_slice[i], coeffIt, m_DerivativeOperator);
+      dx *= this->m_ScaleCoefficients[i];
+      dx_avg += itk::Math::sqr(dx);
+    }
+    dx_avg = std::sqrt(dx_avg);
+
+    // Laplacian
+    dx2_avg = m_InnerProduct(coeffIt, m_LaplacianOperator);
+
+    c.push_back(ComputeDiffusionCoefficient(coeffIt.GetPixel(m_Center), dx_avg, dx2_avg));
+    ++coeffIt;
+  }
+  // Check that we got values for c[i,j]; c[i,j+1]; c[i+1,j]; and c[i+1,j+1]
+  // TODO can cache or remove last diffusion term as c[i+1,j+1] is never used
```

Comment:
The solution for best performance would be to cache diffusion terms as we go as we need a 2x2 local neighborhood to get each pixel delta. However this would require the `Function` object to have foreknowledge of the size of the image buffer rather than operating only on the pixel neighborhoods it is given by its accompanying `ImageFilter`.

Would adding a cache to the function object violate the `FiniteDifference` filter/function design, or is this an acceptable extension?

Reply:
Maybe this filter could have been implemented outside of FiniteDifference hierarchy, so it doesn't have to recompute those terms too much. Recomputing it on work-unit boundaries is hard to avoid, but that would be negligible.

Since you already implemented it this way, perhaps just remove `c[i+1,j+1]`?

Reply:
Lesson learned! I will abbreviate the loop to calculate only the first three terms (`c[i,j]`, `c[i,j+1]`, `c[i+1,j]`).